### DISCUSSION TITLE
What is the Mirth suggested way of adding JDBC drivers that are not part of the defaults?

### DISCUSSION THREAD

**Discussion by LeviLong01** [nextgenhealthcare/connect#4674] [Q&A]
I am working on connecting to a Snowflake database and have successfully done so in Mirth by adding the Snowflake JDBC driver to the `server-lib/database` directory for the Mirth setup.

Is this the intended way Mirth had in mind for people to extend the supported databases on their own? If I add it to `custom-lib` directory it'll throw an error when running saying it can't find the driver.

**Comment by jonbartels** [comment]
MC 3.8.0 added the DB driver code to the config database. It used to be in dbdrivers.xml. What version of MC are you using? The feature is shown on page 316 at https://www.nextgen.com/-/media/files/nextgen-connect/nextgen-connect-311-user-guide.pdf shows how to add this to the list of DBs for the DB connectors, though it doesn't say where to put the drivers.

`server-lib` seems appropriate because that would make the driver available across all of MC on its classpath. That is in contrast with `custom-lib` and Resource Directories which are only loaded for channels that depend on those resources. Also note that `custom-lib` is deprecated and should almost never be used since Resource Directories is a replacement feature with better control.

**Reply by LeviLong01** [reply]
Right now this is tested in Mirth 3.11.

I did setup the `dbdrivers.xml` for if I want to do a Database Reader/Writer, but it'll throw an error if the .jar of the driver is not placed in `server-lib/database`. It doesn't say it needs to be there, but I came to this conclusion of a fix based on this [GitHub Issue](https://github.com/nextgenhealthcare/connect/issues/3538#issuecomment-626957163).

What is a `Resource Directory` for the Mirth setup and is that different than just having a directory named `custom-lib`?

I didn't realize that the `custom-lib` idea was being deprecated.

**Reply by jonbartels** [reply]
A `Resource Directory` works like `custom-lib` in that it is where you should put custom libraries for channels. The advantage of `Resource Directory` is that each directory is separate and each channel can depend on one or more directories. This lets you do things like have two versions of a JAR in MC and have two separate channels use the different versions - you cannot do that if they're all lumped in `custom-lib`.

`Resource Directory` is what one should try FIRST when adding custom libraries or drivers to MC.

Again, in the case of DB drivers to be used everywhere in MC `server-lib/database` is actually the correct spot. That's where DB drivers that are used in many places should be put.

If you are running MC 3.11 you should use the GUI to set up the custom drivers. The release notes for 3.8 say that `dbdrivers.xml` is read "on upgrade" so MC may only ever read that file once when you upgrade from 3.7.x to 3.8 or newer. http://www.mirthcorp.com/community/wiki/display/mirth/Mirth+Connect+3.8.0+-+What%27s+New Since you are running 3.11 use the GUI setup tool to set up the driver you added to `server-lib`.

I also know you hang around Slack. Pop over there if you want some live help. I'm due for a coffee break anyway!

**Reply by tonygermano** [reply]
`custom-lib` is a carry-over from before resource directories existed. It is set up as a default resource on new installs, but it also has special properties that other resource directories don't have.

Ideally, you should put the driver in a resource directory named for the driver, and then you assign it to only the channels that need to use it.

All that being said, I would have expected custom-lib to work. Was mirth already running when you added the driver to the folder? If so, you might have needed to go to the Resources tab in Settings and "reloaded" the resource to make it visible.

**Reply by LeviLong01** [reply]
Got side tracked the past few weeks so I need to hop back in that Slack group @jonbartels!

No it was being restarted each time. @tonygermano

**Reply by tonygermano** [reply]
I would definitely try a `Resource Directory` as that should be sufficient for a driver that is only utilized by channel code. `server-lib` isn't typically where user supplied libraries should go. Part of the confusion with the microsoft driver was that people were trying to use it for the backend database, so it needed to be loaded much earlier than a driver used by a channel would be.

**Reply by tonygermano** [reply]
Maybe related to this issue https://github.com/nextgenhealthcare/connect/issues/4587?

**Reply by someguyyouneverheardof** [reply]
> MC 3.8.0 added the DB driver code to the config database. It used to be in dbdrivers.xml.

From my very limited experience, the GUI editing tool is directly manipulating the dbdrivers.xml file.

**Reply by tonygermano** [reply]
> From my very limited experience, the GUI editing tool is directly manipulating the dbdrivers.xml file.

In upgrades, the dbdrivers.xml file is used to seed the configuration in the database, and then it is no longer used.

**Reply by someguyyouneverheardof** [reply]
Ah, OK. Interesting that the file exists as part of a clean install of 3.11 but I guess it doesn't really hurt.

**Reply by tonygermano** [reply]
@someguyyouneverheardof https://forums.mirthproject.io/forum/mirth-connect/support/18442-dbdrivers-xml-and-3-8-0

**Comment by tonygermano** [comment]
It's always been a complaint of mine that it's never really been spelled out how to do this correctly. It's been less of an issue since the microsoft driver is included in recent versions of mirth as that was the one people were most often trying to use.

How are you trying to use the driver? With one of the database connectors or in javascript? If in javascript, what does your code look like?

**Reply by LeviLong01** [reply]
I've been working with their support to get an answer on if this is the preferred way and I am still not getting a straight answer from them. But based on @jonbartels answer that is the correct location to place it.

JavaScript that would perform the queries, but it is calling to a custom Java package loaded into Mirth for some pre-requisite steps. The Java package has the dependency in a shaded jar (thanks for that tip a while back) so it'll work without it in the `server-lib/database` directory, but when establishing a `DatabaseConnection` object is where it fails if that .jar is not in that directory.
